pkg/importer/spanner/sysl.arrai

```diff
@@ -0,0 +1,129 @@
+
+# attributePatterns generates the patterns for an attribute.
+let attributePatterns = \entity \attr
+    let options = cond {
+        attr.options:
+            let [k, v, ...] = //seq.split('=', attr.options);
+            $`${k}="${v}"`,
+        };
+    let pk = cond {
+        (entity.primary_key where //seq.contains(attr.name, .)):
+            $`~pk${cond {sortingOrder(entity, attr): $`, ${sortingOrder(entity, attr)}`}}`
+    };
+    let fk = cond {
+        (entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_attribute orderby .)) where .:
+            '~fk'
+    };
+    let length = cond {
+        attr.length = 'MAX': '~max',
+    };
+    let hexPrefix = cond {
+        //seq.has_prefix('0x', attr.length): '~hex',
+    };
+    let byteLength = cond {
+        attr.type = 'bytes' && attr.length > 0 && attr.length != 'MAX': $`length="${attr.length}"`
+    };
+    //seq.join(', ', ([options, byteLength, pk, fk, length, hexPrefix] where .@item != ''));
+
+# entityPatterns generates the patterns for an entity.
+let entityPatterns = \entity \model
+    let pk = cond {
+        entity.primary_key count > 1 && compareColumnOrder(entity):
+            $`primary_key="${entity.primary_key orderby .::, }"`,
+    };
+    let cluster = cond {
+        entity.cluster:
+            //seq.join('', entity.cluster >>
+            $`interleave_in_parent="${.interleaved_in}", interleave_on_delete="${.on_delete}"`),
+    };
+    let fk = cond {
+        entity.foreign_keys:
+            $`foreign_keys=[${//seq.join(', ',(entity.foreign_keys =>
+                \keys $`["constraint:${keys.constraint_name}","columns:${//seq.join(', ', (keys.foreign_keys => .attribute) orderby .)}"]`) orderby .)}]`,
+    };
+    let indices = cond {
+        (model.indexes where .table_name = entity.name):
+            $`indexes=[${//seq.join(', ', ((model.indexes orderby .) where .@item.table_name = entity.name) >>
+                $`"name:${.table_name}","unique:${.unique}","null_filtered:${.nullfiltered}","key_parts:${//seq.join(', ', .key_part orderby .)}${cond {.storing_col: $`","storing:${//seq.join(', ',.storing_col)}`}}${cond {.interleaved_table: $`","interleave_in:${.interleaved_table}`}}`)}"]`,
+    };
+    //seq.join(', ', ([pk, cluster, fk, indices] where .@item != ''));
+
+# appendEntityPatterns returns a pattern for the entity patterns together.
+let appendEntityPatterns = \entity \model cond { hasEntityPattern(entity, model): $`[${entityPatterns(entity, model)}]`};
+
+# appendAttributePatterns returns a pattern for the attribute patterns together.
+let appendAttributePatterns = \entity \attr cond { hasAttributePattern(entity, attr): $`[${attributePatterns(entity, attr)}]`};
+
+# typeInfo generates the type info for an attribute.
+let typeInfo = \entity \attr \type \isArray
+    let fkAttr = entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_attribute orderby .) where .;
```

Comment:
Nit: I've seen `entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) => .reference_attribute orderby .) where .;` four times in this file, and it makes whatever code it's in look more complex. Candidate for refactoring into a function?

Reply:
refactored